homeassistant/components/rointe/climate.py

```diff
@@ -0,0 +1,227 @@
+            return HVACMode.AUTO
+
+        return HVACMode.HEAT
+
+    @property
+    def hvac_action(self) -> HVACAction:
+        """Return the current HVAC action."""
+
+        # Special mode for AUTO mode and waiting for schedule to activate.
+        if (
+            self._radiator.mode == RointeOperationMode.AUTO.value
+            and self._radiator.preset == HVACMode.OFF
+        ):
+            return HVACAction.IDLE
+
+        # Forced to off, either on Manual or Auto mode.
+        if not self._radiator.power:
+            return HVACAction.OFF
+
+        # Otherwise, it's heating.
+        return HVACAction.HEATING
+
+    @property
+    def preset_mode(self) -> str | None:
+        """Convert the device's preset to HA preset modes."""
+
+        # Also captures "none" (man mode, temperature outside presets)
+        return ROINTE_HASS_MAP.get(self._radiator.preset, None)
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set new target temperature."""
+
+        target_temperature = kwargs["temperature"]
+
+        LOGGER.debug("Setting temperature to %s", target_temperature)
+
+        if not RADIATOR_TEMP_MIN <= target_temperature <= RADIATOR_TEMP_MAX:
+            raise ValueError(
+                f"Invalid set_temperature value (must be in range {RADIATOR_TEMP_MIN}, {RADIATOR_TEMP_MAX}): {target_temperature}"
+            )
+
+        # Round to the nearest half value.
+        rounded_temp = round(target_temperature * 2) / 2
+
+        if not await self.device_manager.send_command(
+            self._radiator, RointeCommand.SET_TEMP, rounded_temp
+        ):
+            raise HomeAssistantError(
+                f"Failed to set temperature for {self._radiator.name}"
+            )
+
+        await self._signal_thermostat_update()
+
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set new target hvac mode."""
+
+        LOGGER.debug("Setting HVAC mode to %s", hvac_mode)
+
+        if not await self.device_manager.send_command(
+            self._radiator, RointeCommand.SET_HVAC_MODE, hvac_mode
+        ):
+            raise HomeAssistantError(
+                f"Failed to set HVAC mode for {self._radiator.name}"
+            )
+
+        await self._signal_thermostat_update()
+
+    async def async_set_preset_mode(self, preset_mode: str) -> None:
+        """Set new target preset mode."""
+        LOGGER.debug("Setting preset mode: %s", preset_mode)
+
+        if not await self.device_manager.send_command(
+            self._radiator, RointeCommand.SET_PRESET, preset_mode
+        ):
+            raise HomeAssistantError(
+                f"Failed to set HVAC preset for {self._radiator.name}"
+            )
+
+        await self._signal_thermostat_update()
+
+    async def _signal_thermostat_update(self):
+        """Signal a radiator change."""
+
+        # Update the data
+        await self.coordinator.async_refresh()
+        self.async_write_ha_state()
```

Comment:
I don't think you need to write HA state here, because the moment that the coordinator updates, `_handle_coordinator_callback` is called, which also writes ha state, so this now creates 2 state writes instead of 1